File: com/jseidemann/adventofcode/day2/Day2Pt1.java
package com.jseidemann.adventofcode.day2;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

public class Day2Pt1 {

    public static void main(String[] args) throws IOException {
        List<String> inputStrings = Files.readAllLines(Paths.get("com/jseidemann/adventofcode/day2/adventofcode_puzzle2_input.txt"));

        //@formatter:off
        int total = inputStrings.stream()
                .map(Round::new)
                .mapToInt(Round::getScore)
                .sum();
        //@formatter:on

        System.out.println("I got " + total + " points");
    }

    static class Round {
        OppPlay oppPlay;
        MyPlay myPlay;

        public Round(String input) {
            this.oppPlay = new OppPlay(input.substring(0, 1));
            this.myPlay = MyPlay.valueOf(input.substring(2, 3));
        }

        public Integer getScore() {
            System.out.println("I played " + myPlay + ". They played " + oppPlay.choice + ". So I got " + (myPlay.value + oppPlay.calculateWinner(myPlay)) + " points.");
            return myPlay.value + oppPlay.calculateWinner(myPlay);
        }
    }

    enum MyPlay {
        X(1), Y(2), Z(3);

        Integer value;

        MyPlay(Integer val) {
            this.value = val;
        }
    }

    static class OppPlay {
        String choice;

        public OppPlay(String choice) {
            this.choice = choice;
        }

        Integer calculateWinner(MyPlay myPlay) {
            if (choice.equals("A")) {
                switch (myPlay) {
                    case X: return 3;
                    case Y: return 6;
                    case Z: return 0;
                    default: throw new RuntimeException();
                }
            } else if (choice.equals("B")) {
                switch (myPlay) {
                    case X: return 0;
                    case Y: return 3;
                    case Z: return 6;
                    default: throw new RuntimeException();
                }
            } else {
                switch (myPlay) {
                    case X: return 6;
                    case Y: return 0;
                    case Z: return 3;
                    default: throw new RuntimeException();
                }
            }
        }
    }
}

File: com/jseidemann/adventofcode/day2/Day2Pt1.kt
package com.jseidemann.adventofcode.day2

import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths

object Day2Pt1 {
    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val inputStrings = Files.readAllLines(Paths.get("com/jseidemann/adventofcode/day2/adventofcode_puzzle2_input.txt"))

        //@formatter:off
        val total = inputStrings.stream()
                .map(::Round)
                .mapToInt(Round::calculateScore)
                .sum()
        //@formatter:on
        println("I got $total points")
    }

    internal class Round(input: String) {
        var oppPlay: OppPlay = OppPlay(input.substring(0, 1))
        var myPlay: MyPlay = MyPlay.valueOf(input.substring(2, 3))

        fun calculateScore(): Int {
            println("I played " + myPlay + ". They played " + oppPlay.choice + ". So I got " + (myPlay.value + oppPlay.calculateWinner(myPlay)) + " points.")
            return myPlay.value + oppPlay.calculateWinner(myPlay)
        }

    }

    internal enum class MyPlay(var value: Int) {
        X(1), Y(2), Z(3);
    }

    internal class OppPlay(var choice: String) {
        fun calculateWinner(myPlay: MyPlay?): Int {
            return when (choice) {
                "A" -> {
                    when (myPlay) {
                        MyPlay.X -> 3
                        MyPlay.Y -> 6
                        MyPlay.Z -> 0
                        else -> throw RuntimeException()
                    }
                }
                "B" -> {
                    when (myPlay) {
                        MyPlay.X -> 0
                        MyPlay.Y -> 3
                        MyPlay.Z -> 6
                        else -> throw RuntimeException()
                    }
                }
                "C" -> {
                    when (myPlay) {
                        MyPlay.X -> 6
                        MyPlay.Y -> 0
                        MyPlay.Z -> 3
                        else -> throw RuntimeException()
                    }
                }
                else -> throw RuntimeException()
            }
        }
    }
}